src/Data/AIG/CompactGraph.hs

```diff
@@ -0,0 +1,444 @@
+{- |
+Module      : Data.AIG.CompactGraph
+Copyright   : (c) Galois, Inc. 2021
+License     : BSD3
+Maintainer  : atomb@galois.com
+Stability   : experimental
+Portability : portable
+
+A pure Haskell implementation of the IsAIG class with support for AIGER
+and CNF file creation.
+-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Data.AIG.CompactGraph
+  ( CompactGraph
+  , CompactLit
+  , CompactNetwork
+  , compactProxy
+  , newCompactGraph
+  ) where
+
+import Control.Monad (forM_, replicateM)
+import Data.Binary.Get
+import qualified Data.Binary.Parser as BP
+import Data.Bits (shiftL, shiftR, (.&.), (.|.), xor, testBit)
+import Data.IORef (IORef, newIORef, modifyIORef', readIORef, writeIORef)
+import Data.List (elemIndex, intersperse)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.ByteString.Builder as BS
+import Data.Word (Word32)
+import System.IO (Handle, withFile, IOMode(..))
+
+import Data.AIG.Interface hiding (xor)
+
+newtype Var = Var Word32
+  deriving (Eq, Ord, Show, Enum)
+
+nextVar :: Var -> Var
+nextVar (Var v) = Var (v + 1)
+
+------------------------------------------------------------------
+-- | A compact "Graph" datastructure similar to the AIGER format.
+data CompactGraph s =
+  CompactGraph {
+    maxVar  :: IORef Var
+    -- ^ The largest variable ID used so far.
+  , inputs  :: IORef [Var]
+    -- ^ Inputs, in reverse order!
+  , andMap  :: IORef (Map Var (CompactLit s, CompactLit s))
+    -- ^ A map from and gate variables to their input literals.
+  }
+
+------------------------------------------------------------------
+-- | A literal in a CompactGraph.
+newtype CompactLit s = CompactLit Word32
```

Comment:
I think it's probably worth writing some documentation regarding the relationship between literals and variables.  It's a bit subtle if you haven't seen it before.

Reply:
Added.